File: src/sequence/core/SqVars.java
package sequence.core;

import sequence.world.entities.SpreadPointBulletType;
import arc.Events;
import arc.files.Fi;
import arc.func.Cons;
import arc.math.Interp;
import arc.struct.IntMap;
import arc.struct.Seq;
import mindustry.Vars;
import mindustry.game.EventType;
import mindustry.mod.ClassMap;
import org.jetbrains.annotations.NotNull;

public class SqVars {
    private static final int MIN_PRIORITY = Integer.MAX_VALUE;

    public static final Interp pow7 = new Interp.Pow(7);
    public static final Object[] emptyObjArr = {};
    public static final float revTimerUnit = 20;
    public static Fi unzipDirectory;
    public static final IntMap<Seq<Processor>> processors = new IntMap<>();

    static {
        unzipDirectory = Vars.dataDirectory.child("unzip/");
        Events.run(EventType.Trigger.update, () -> allProcessors(Processor::run));
    }

    public static void addProcessor(Runnable runnable, int priority) {
        Processor processor = new Processor(runnable, priority);
        processors.get(priority, Seq::new).add(processor);
    }

    public static void addProcessor(Runnable runnable) {
        addProcessor(runnable, MIN_PRIORITY);
    }

    public static void allProcessors(Cons<Processor> action) {
        for (IntMap.Entry<Seq<Processor>> entry : processors) {
            for (Processor p : entry.value) {
                action.get(p);
            }
        }
    }

    public static final class Processor implements Comparable<Processor> {
        public Runnable runnable;
        public int priority = 0;

        public Processor(Runnable runnable) {
            this.runnable = runnable;
        }

        public Processor(Runnable runnable, int priority) {
            this.runnable = runnable;
            this.priority = priority;
        }

        public void run() {
            runnable.run();
        }

        @Override
        public int compareTo(@NotNull Processor o) {
            return Integer.compare(priority, o.priority);
        }
    }
}

File: src/sequence/core/SqVars.kt
package sequence.core

import arc.Events
import arc.files.Fi
import arc.func.Cons
import arc.math.Interp
import arc.struct.IntMap
import arc.struct.Seq
import mindustry.Vars
import mindustry.game.EventType

object SqVars {
    //    val sha3_512 = MessageDigest.getInstance("SHA3-512")
    val pow7: Interp = Interp.Pow(7)
    val emptyObjArr = arrayOf<Any>()
    const val revTimerUnit = 20f
    val processors = IntMap<Seq<Processor>>()
    private const val MIN_PRIORITY = Int.MAX_VALUE
    val unzipDirectory: Fi by lazy { Vars.dataDirectory.child("unzip/") }

    init {
        Events.run(EventType.Trigger.update) { allProcessors { it.run() } }
    }

    @JvmOverloads
    fun addProcessor(runnable: Runnable, priority: Int = MIN_PRIORITY) {
        val processor = Processor(runnable, priority)
        processors[priority, { Seq() }].add(processor)
    }

    fun allProcessors(action: Cons<Processor>) {
        for (entry in processors) {
            for (p in entry.value) {
                action[p]
            }
        }
    }

    class Processor @JvmOverloads constructor(var runnable: Runnable, var priority: Int = 0) : Comparable<Processor> {
        fun run() {
            runnable.run()
        }

        override fun compareTo(other: Processor): Int {
            return priority.compareTo(other.priority)
        }
    }
}
